src/consensus/merkle.cpp

```diff
@@ -185,3 +263,296 @@ std::vector<uint256> BlockMerkleBranch(const CBlock& block, uint32_t position)
+    m_proof.m_skip.insert(m_proof.m_skip.end(), left.m_proof.m_skip.begin(), left.m_proof.m_skip.end());
+    m_verify.insert(m_verify.end(), left.m_verify.begin(), left.m_verify.end());
+
+    if (right.m_proof.m_path.empty())
+        m_proof.m_path.front().SetRight(right.m_proof.m_skip.empty()? MerkleLink::VERIFY: MerkleLink::SKIP);
+    m_proof.m_path.insert(m_proof.m_path.end(), right.m_proof.m_path.begin(), right.m_proof.m_path.end());
+    m_proof.m_skip.insert(m_proof.m_skip.end(), right.m_proof.m_skip.begin(), right.m_proof.m_skip.end());
+    m_verify.insert(m_verify.end(), right.m_verify.begin(), right.m_verify.end());
+}
+
+void MerkleTree::clear() noexcept
+{
+    m_proof.clear();
+    m_verify.clear();
+}
+
+void swap(MerkleTree& lhs, MerkleTree& rhs)
+{
+    using std::swap;
+    swap(lhs.m_proof, rhs.m_proof);
+    swap(lhs.m_verify, rhs.m_verify);
+}
+
+uint256 MerkleTree::GetHash(bool* invalid) const
+{
+    std::vector<std::pair<bool, uint256> > stack(2);
+    auto verify_pos = m_verify.begin();
+    auto verify_last = m_verify.end();
+    auto skip_pos = m_proof.m_skip.begin();
+    auto skip_last = m_proof.m_skip.end();
+
+    auto visitor = [&stack, &verify_pos, &verify_last, &skip_pos, &skip_last](std::size_t depth, MerkleLink value, bool right) -> bool
+    {
+        const uint256* new_hash = nullptr;
+        switch(value) {
+            case MerkleLink::DESCEND:
+                stack.emplace_back();
+                return false;
+
+            case MerkleLink::VERIFY:
+                if (verify_pos == verify_last) // read past end of verify hashes list
+                    return true;
+                new_hash = &(verify_pos++)[0];
+                break;
+
+            case MerkleLink::SKIP:
+                if (skip_pos == skip_last) // read past end of skip hashes list
+                    return true;
+                new_hash = &(skip_pos++)[0];
+                break;
+        }
+
+        uint256 tmp;
+        while (stack.back().first) {
+            MerkleHash_Sha256Midstate(tmp, stack.back().second, *new_hash);
+            new_hash = &tmp;
+            stack.pop_back();
+        }
+
+        stack.back().first = true;
+        stack.back().second = *new_hash;
+        return false;
+    };
+
+    // As a special case, an empty proof with no verify hashes results
+    // in the unsalted hash of empty string. Although this requires
+    // extra work in this implementation to support, it provides
+    // continuous semantics to the meaning of the MERKLEBLOCKVERIFY
+    // opcode, which might potentially reduce the number of code paths
+    // in some scripts.
+    if (m_proof.m_path.empty() && m_verify.empty() && m_proof.m_skip.empty()) {
+        if (invalid) {
+            *invalid = false;
+        }
+        return CHashWriter(SER_GETHASH, PROTOCOL_VERSION).GetHash();
+    }
+
+    // Except for the special case of a 0-node, 0-verify, 0-skip tree,
```

Comment:
I think this is the reason for allowing invalid=true in MBV. Maybe it should be fixed by setting *invalid to the logical OR of the various sizes referenced.

Reply:
All tests passed, though. I'm pretty sure you are testing for that case, no?